ethpm_cli/scraper.py

```diff
@@ -37,24 +43,57 @@ def scrape(w3: Web3, ethpm_dir: Path, start_block: int = 0) -> int:
             f"instance with latest block number of {latest_block}."
         )
 
-    ethpmdir_block = max(get_scraped_blocks(chain_data_path))
-    active_block = start_block if start_block else ethpmdir_block
-    for block in range(active_block, latest_block):
-        try:
-            validate_unscraped_block(block, chain_data_path)
-        except BlockAlreadyScrapedError:
-            pass
+    if start_block == 0:
+        active_block = get_ethpm_birth_block(w3, 0, latest_block)
+    else:
+        active_block = start_block
+
+    logger.info("Scraping from block %d.", active_block)
+    for from_block in range(active_block, latest_block, BLOCK_INTERVAL):
+        if (from_block + BLOCK_INTERVAL) > latest_block:
+            to_block = latest_block
         else:
-            scraped_manifests = scrape_block_for_manifests(w3, block)
-            update_chain_data(chain_data_path, block, scraped_manifests)
+            to_block = from_block + BLOCK_INTERVAL
+
+        if block_range_needs_scraping(from_block, to_block, chain_data_path):
+            scraped_manifests = scrape_block_range_for_manifests(
+                w3, from_block, to_block
+            )
+            update_chain_data(chain_data_path, from_block, to_block, scraped_manifests)
             write_ipfs_uris_to_disk(ethpm_dir, scraped_manifests)
+        else:
+            logger.info("Block range: %d - %d already scraped.", from_block, to_block)
+
     return latest_block
 
 
-def validate_unscraped_block(block: int, chain_data_path: Path) -> None:
+def get_ethpm_birth_block(w3: Web3, from_block: int, to_block: int) -> int:
+    """
+    Returns the closest block found before the EthPM VersionRelease event birthday.
+    """
+    version_release_date = datetime.fromtimestamp(VERSION_RELEASE_TIMESTAMP)
+    from_date = datetime.fromtimestamp(w3.eth.getBlock(from_block)["timestamp"])
+    delta = version_release_date - from_date
+
+    if delta.days == 0 and from_date < version_release_date:
+        return from_block
+
+    elif from_date < version_release_date:
+        updated_block = int((from_block + to_block) / 2)
+        return get_ethpm_birth_block(w3, updated_block, to_block)
```

Comment:
Yeah, no problem. I'm starting to sense that you don't think recursive solutions are the right tool to reach for most of the time?

Reply:
Nah, for this the recursion limit is very unlikely to be hit, but that's assuming that the program executes correctly.  I've re-implemented binary searches a bunch of time and plenty of them ended up with bugs that resulted in an infinite loop which would have translated to an error due to stack depth here.   Neither of those are really better necessarily.

Really it is that recursion is a sort of code smell.  

- It costs extra in some ways because python has to allocate more call frames.
- You have to know the bounds of your algorithm to evaluate if stack limit will be a problem.
- other things that probably show up if you google "why is recursion bad"

Maybe the right thing would be *"Recursion is by default the wrong tool, and is only the right tool if it has been closely evaluated in the context of the problem it solves".